src/main/java/org/dependencytrack/persistence/jdbi/ProjectDao.java

```diff
@@ -227,4 +228,39 @@ record ConciseProjectMetricsRow(
              WHERE "UUID" = :projectUuid
             """)
     int deleteProject(@Bind final UUID projectUuid);
+
+    @SqlUpdate("""
+           DELETE
+            FROM "PROJECT"
+            WHERE "PROJECT"."INACTIVE_SINCE" IS NOT NULL
+            AND NOW() - "PROJECT"."INACTIVE_SINCE" > :duration
+            AND "PROJECT"."ID" = any( array( SELECT "ID" FROM "PROJECT" LIMIT :batchSize));
+           """)
+    int deleteInactiveProjectsForRetentionDuration(@Bind final Duration duration, @Bind final int batchSize);
+
+    @SqlUpdate("""
+           DELETE
+            FROM "PROJECT"
+            WHERE "PROJECT"."INACTIVE_SINCE" IS NOT NULL
+            AND "PROJECT"."NAME" = :projectName
+            AND "PROJECT"."ID" NOT IN (
+                SELECT "PROJECT"."ID" 
+                 FROM "PROJECT"
+                 WHERE "PROJECT"."INACTIVE_SINCE" IS NOT NULL
+                 AND "PROJECT"."NAME" = :projectName
+                 ORDER BY "PROJECT"."INACTIVE_SINCE" DESC
```

Comment:
Won't `DESC` ordering not always delete the "youngest" inactive projects first?

Reply:
No, here sorting DESC will give list of youngest inactive projects and that list will be excluded : 
```
AND "PROJECT"."ID" NOT IN (
  SELECT "PROJECT"."ID" 
   FROM "PROJECT"
   WHERE "PROJECT"."INACTIVE_SINCE" IS NOT NULL
   AND "PROJECT"."NAME" = :projectName
   ORDER BY "PROJECT"."INACTIVE_SINCE" DESC
   LIMIT :retentionCadence
  )
```